### PR TITLE
ACNA-882 - Rewriting files during app run is error prone, and leads to unexpected issues

### LINKED ISSUE #275: Rewriting files during app run is error prone, and leads to unexpected issues.
**Describe the bug**
When a developer runs `aio app run --local` we replace some of the local config while the app is being run.
.env gets backed up to .env.app.save
.wskdebug.props.tmp gets written
.vscode/launch.json gets written if it does not exist, or backed up and overwritten if it did.
There are potentially other areas where we do this also.

**To Reproduce**
1. with vscode open, run `aio app run`
1. modify the contents of .env file
1. cmd/ctrl+c in your terminal to stop running
1. your changes to .env are gone

**Expected behavior**
It should be obvious what is happening.
If a file is machine-modified, it should include a header explaining that, and informing the user not to modify it.
There should be no files that are both machine + user modified or there will always be issues.

**Desktop (please complete the following information):**
 - OS: any
 - Browser: any/na
 - Version: @adobe/aio-cli-plugin-app <= 3.4.0

**Additional context**
There are likely different solutions for different files here.
the vscode changes should possibly be written during app:init if they can [ACNA-712] is related.
this should likely be split into separate issues for each affected file.

### PR DESCRIPTION
**Pull request by shazron**
Fixes #275 

## How Has This Been Tested?

- npm test
- manual test on macOS (browser action test, vscode local debugging test)
- manual test on Linux Ubuntu 20 (see workaround https://github.com/adobe/aio-cli-plugin-app/issues/311) (browser action test, vscode local debugging test)
- manual test on Windows 10 (re-fixed https://github.com/adobe/aio-cli-plugin-app/issues/304 via https://github.com/apache/openwhisk/pull/5021)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by shazron** [comment]
Two things that need to be fixed still:
1. local debugging. Setting WSK_CONFIG_FILE, the default root .env still takes precedence. Need to set and write to the .env in `.vscode/launch.json`:
```
"envFile": "${workspaceFolder}/dist/.env",
```
2. in `dist/.env`, need to write these values:
```
OW_APIHOST
OW_NAMESPACE
OW_AUTH
```
These wsk vars take precedence over everything else in wskdebug.

Already tested and they work ok, just need to cleanup some things.

**Comment by shazron** [comment]
Item left: fixing tests

**Comment by shazron** [comment]
On macOS tests are 100% but are failing on Travis CI (Linux and Windows). WIll have to pick this up next week and test on those OSes.

**Comment by shazron** [comment]
Two issues still:
1. runDev.js line 157 branch coverage: it doesn't show on macOS, but on Linux and Windows, so need to iterate and dev on those platforms to get coverage full.
2. Windows has some test failures related to vscode expected configuration not matching

**Comment by purplecabbage** [comment]
I find potentially more confusing.  Now I don't have a .vscode/launch.json and then I do, and then I don't ...
Why can't launch.json be persistent? It does not change between `aio app run` and `aio app run --local`

Also, with .env file now in dist/ ... we have one which is a copy of ../.env when run without --local,  and one with ... can we just keep a local.env and prod.env in the dist/ folder instead?

**Comment by shazron** [comment]
> I find this potentially more confusing. Now I don't have a .vscode/launch.json and then I do, and then I don't ...
> Why can't launch.json be persistent? It does not change between `aio app run` and `aio app run --local`

I didn't change the existing functionality, so this behaviour is consistent with the previous behaviour. Not sure the rationale for this functionality.
 
> Also, with .env file now in dist/ ... we have one which is a copy of ../.env when run without --local, and one with ... can we just keep a local.env and prod.env in the dist/ folder instead?

Not sure I follow exactly. When you run with `--local` it uses `dist/.env`. When run without `--local` it uses your root `.env` and nothing has changed there.

**Comment by shazron** [comment]
Windows 10 manual tests have been done (via a local patch, see Description), and local debugging etc works.

**Review by moritzraho — COMMENTED** [review]
One inline question about the stop log arg.
Didn't tested, just looked at the code for now, but this looks mostly fine to me.

I agree with @purplecabbage about the vscode file, that it should be generated via generator-aio-app (on init), but I think this could be addressed in another PR

About the dist/.env discussion, I guess one potential enhancement would be, **if possible**,  to pass wskdebug vars via environment instead of `dist/.env`, so that we can rename `dist/.env` to `dist/.env.local` (makes the intent clearer for advanced devs looking at `dist/`) and only have it generated on `--local`. But again not sure if it's possible or should be done as part of this PR

Also thanks for the extra refactoring and cleanup work !

**Comment by shazron** [comment]
> I agree with @purplecabbage about the vscode file, that it should be generated via generator-aio-app (on init), but I think this could be addressed in another PR

I left this precisely for another PR, since I saw that comment somewhere about moving to the generator eventually.
 
> About the dist/.env discussion, I guess one potential enhancement would be if **possible** to pass wskdebug vars via environment instead of `dist/.env`, so that we can rename `dist/.env` to `dist/.env.local` (makes the intent clearer for advanced devs looking at `dist/`) and only have it generated on `--local`. But again not sure if it's possible or should be done as part of this PR

We could pass it in as env vars instead of dist/.env, but I'm not sure what the advantage is since dist/.env is functionally equivalent and cleaner. We can rename dist/.env to anything so that is not a problem in renaming it.

File: src/lib/runDev.js [hunk 1]
```diff
@@ -182,158 +185,68 @@ async function runDev (args = [], config, options = {}, log) {
...
-      breakOnLoad: true,
-      sourceMapPathOverrides: {
-        '*': path.join(devConfig.web.distDev, '*')
-      }
-    })
-    debugConfig.compounds.push({
-      name: 'WebAndActions',
-      configurations: ['Web'].concat(actionConfigNames)
-    })
+  try {
```
[the remark this answers is not in the document]
Comment (shazron) [reply src/lib/runDev.js:236]: `if (config.app.hasBackend && fetchLogs)` block should already handle this.

File: src/lib/cleanup.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const aioLogger = require('@adobe/aio-lib-core-logging')('@adobe/aio-cli-plugin-app:cleanup', { provider: 'debug' })
+
+/** @private */
+class Cleanup {
```
[the remark this answers is not in the document]
Comment (shazron) [reply src/lib/cleanup.js:17]: 👍🏼

File: src/lib/runDev.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (moritzraho) [reply src/lib/runDev.js:236]: yup thanks got confused and missed the fetchLogs option. Thanks for clarifying

**Comment by moritzraho** [comment]
> We could pass it in as env vars instead of dist/.env, but I'm not sure what the advantage is since dist/.env is functionally equivalent and cleaner. We can rename dist/.env to anything so that is not a problem in renaming it.

I meant to limit the cases where we generate the extra file, but ok might be more confusing to have multiple means of setting env variables indeed. Maybe we can add some comments in the generated .env instead ?

**Comment by shazron** [comment]
Ok I think I get what you mean Moritz.

1. for `--local` generate the `dist/.env.local` with comments
2. if `--local` not present, use the root `.env`

I will keep the existing `.vscode/launch.json` functionality since it was the same behaviour before (most likely so we don't potentially clobber their changes).
I will leave the vscode file generation to the generator for another PR.

**Comment by moritzraho** [comment]
ok agree to do this in incremental steps 🙂

**Review by moritzraho — APPROVED** [review]

**Comment by shazron** [comment]
Travis-CI builds have stopped on this org. Migrating to Github Actions in another PR, and when that is integrated, we can merge it into this PR.

**Pull request merged**